### ISSUE TITLE
Provide a way to specify headers from the message

### ISSUE THREAD

**Issue by elyscape** [giraffi/fluent-plugin-amqp#35]
It would be nice if there were a way to set headers for a message based on the contents of the message. For example, if we have messages that look like the following (taken from [the parser plugin docs](http://docs.fluentd.org/articles/parser-plugin-overview):
```
2013-03-03 14:27:33 +0900 zimbra.mailbox: {"thread":"main","level":"INFO","message":"  Main - Start"}
2013-03-03 14:27:33 +0900 zimbra.mailbox: {"thread":"main","level":"ERROR","message":" Main - Exception\njavax.management.RuntimeErrorException: null\n    at Main.main(Main.java:16) ~[bin/:na]"}
2013-03-03 14:27:33 +0900 zimbra.mailbox: {"thread":"main","level":"INFO","message":"  Main - End"}
```
It would be nice to be able to set a header on the message when sent to RabbitMQ whose value is the that of the `level` field, maybe by doing something like this:
```
<match **>
  type amqp
  exchange amq.headers
  host amqp.example.com
  port 5672
  vhost /
  user guest
  pass guest
  format json
  <header>
    name LogLevel
    source level
    default "INFO"
  </header>
</match>
```
This would make it possible to use Headers exchanges properly.

**Comment by sawanoboly** [comment]
hmm..
Perhaps, I think that it can be processed beforehand by using a filter plugins.

**Comment by elyscape** [comment]
The tag can be modified using filter plugins, but there's no way to set custom headers. The code that assigns headers only supports setting a header to contain the timestamp and a header to contain the message tag.

[headers]: https://github.com/giraffi/fluent-plugin-amqp/blob/v0.10.1/lib/fluent/plugin/out_amqp.rb#L133-L138

**Comment by sawanoboly** [comment]
Ah, I understand. 
What do you think about this function? @warmfusion

**Comment by warmfusion** [comment]
I like it; useful for filtering and shaping events through exchanges. Not had to use headers much yet, so it might need some investigation and testing

**Comment by warmfusion** [comment]
@sawanoboly  - I've implemented this in #40 and as the updated README indicates its used as follows;

#### headers

It is possible to specify message headers based on the content of the incoming
message, or as a fixed default value as shown below;

```
<matcher ...>
...

  <header>
    name LogLevel
    source level
    default "INFO"
  </header>
  <header>
    name SourceHost
    default my.example.com
  </header>
  <header>
    name CorrelationID
    source x-request-id
  </header>
...
</matcher>
```


The header elements may be set multiple times for multiple additional headers
to be included on any given message.

* If source is omitted, the header will _always_ be set to the default value
* If default is omitted the header will only be set if the source is found
* Note that values from nested json keys is not supported

**Comment by sawanoboly** [comment]
released as v0.11.0 thanks,
